src/main/java/mate/academy/AsyncRequestProcessor.java

```diff
@@ -1,16 +1,40 @@
 package mate.academy;
 
+import java.util.Map;
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.Executor;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.TimeUnit;
 
 public class AsyncRequestProcessor {
+    public static final String DETAILS_HEAD = "Details for ";
+    private final Map<String, UserData> cache = new ConcurrentHashMap<>();
     private final Executor executor;
 
     public AsyncRequestProcessor(Executor executor) {
         this.executor = executor;
     }
 
     public CompletableFuture<UserData> processRequest(String userId) {
-        return null;
+        return CompletableFuture.supplyAsync(() -> getUserData(userId), executor);
+    }
+
+    private UserData getUserData(String userId) {
+        try {
+            long timeout = ThreadLocalRandom.current().nextLong(2000);
+            TimeUnit.MILLISECONDS.sleep(timeout);
+        } catch (InterruptedException e) {
+            System.out.println("Interrupted for " + userId);
+            Thread.currentThread().interrupt();
+        }
+        UserData cachedUserData = cache.get(userId);
+        if (cachedUserData != null) {
+            return cachedUserData;
+        }
```

Comment:
what is the point of cache if you are sleeping the thread without checking the cache before that?
* we need sleep here to simulate some DB access & fetch time
* cache is for storing the data that we have already fetched earlier in memory to not waste time for fetching it once more

So let's move the thread sleep and the `UserData` object build to some separate method with only that, just like we would have with our repository, some `getUserDataById(userId)` and use it somewhere in our logic